src/program/latex/diagram/mod.rs

```diff
@@ -0,0 +1,176 @@
+#[derive(Clone, Debug, Default)]
+pub(super) struct Diagram {
+    /// customizes how the diagram renders the circuit
+    pub(crate) settings: RenderSettings,
+    /// a BTreeMap of wires with the name of the wire as the key
+    pub(crate) circuit: BTreeMap<u64, Box<Wire>>,
+}
+
+impl Diagram {
+    /// Compares qubits from a single instruction associated with a column on
+    /// the circuit to all of the qubits used in the quil program. If a qubit
+    /// from the quil program is not found in the qubits in the single
+    /// instruction line, then an empty slot is added to that column on the
+    /// qubit wire of the circuit indicating a "do nothing" at that column.
+    ///
+    /// # Arguments
+    /// `qubits` - exposes the qubits used in the Program
+    /// `instruction` - exposes the qubits in a single Instruction
+    pub(crate) fn apply_empty(&mut self, qubits: &HashSet<Qubit>, gate: &Gate) {
+        qubits
+            .difference(&gate.qubits.iter().cloned().collect())
+            .filter_map(|q| match q {
+                Qubit::Fixed(index) => Some(index),
+                _ => None,
+            })
+            .for_each(|index| {
+                if let Some(wire) = self.circuit.get_mut(index) {
+                    // increment the wire column
+                    wire.column += 1;
+                    wire.set_empty()
+                }
+            });
+    }
+
+    /// Applies a gate from an instruction to the wires on the circuit
+    /// associate with the qubits from the gate. If the gate name matches a
+    /// composite gate, then the composite gate is applied to the circuit,
+    /// otherwise, the original gate is applied to the circuit.
+    ///
+    /// # Arguments
+    /// `gate` - the Gate of the Instruction from `to_latex`.
+    pub(crate) fn apply_gate(&mut self, gate: &Gate) -> Result<(), LatexGenError> {
+        // for each fixed qubit in the gate
+        for qubit in &gate.qubits {
+            if let Qubit::Fixed(qubit) = qubit {
+                if let Some(wire) = self.circuit.get_mut(qubit) {
+                    // increment the wire column
+                    wire.column += 1;
+
+                    // set modifiers at this column for all qubits
+                    wire.set_daggers(&gate.modifiers)?;
+
+                    // set parameters at this column for all qubits
+                    for expression in &gate.parameters {
+                        wire.set_param(expression, self.settings.texify_numerical_constants);
+                    }
+                }
+            }
+        }
+
+        // get display of gate name from composite gate or original gate
+        let canonical_gate = CompositeGate::from_str(&gate.name)
+            .map(|g| g.to_string())
+            .unwrap_or(gate.name.clone());
+
+        // get the names of the qubits in the circuit before circuit is borrowed as mutable
+        let circuit_qubits: Vec<u64> = self.circuit.keys().cloned().collect();
```

Comment:
I can see how this works with qubits that are imputed but not the other way. Lets say we have a range of 2 for qubits [2,5] and 5 controls 2, so it should be 5{-1} (5 is a distance of -1 from 2). So if I use the range to find this distance, then 5-2 = 3, and 2-2 = 0, 0-3 = -3. Even if I make the adjustment to range -1, so 1, then I still end up with -3, 5-1 = 4, 2-1 = 1, 1-4 = -3. 

I get that we can use the range to shift the qubits, but this works well when the qubits are imputed. Otherwise it doesn't shift the qubits correctly as it assumes they're imputed. I'm not seeing a very straight forward way to use this. I find is easier to locate these qubits in the full circuit by index and just subtract the index. How are you using this in your version?